src/p2p.rs

```diff
@@ -91,43 +91,66 @@ impl Connection {
     where
         B: IntoIterator<Item = BlockHash>,
-        F: FnMut(BlockHash, SerBlock),
+        F: Fn(BlockHash, SerBlock) -> R + Send + Sync,
+        R: Send + Sync,
     {
         self.blocks_duration.observe_duration("total", || {
             let blockhashes: Vec<BlockHash> = blockhashes.into_iter().collect();
+            let blockhashes_len = blockhashes.len();
             if blockhashes.is_empty() {
-                return Ok(());
+                return Ok(vec![]);
             }
             self.blocks_duration.observe_duration("request", || {
                 debug!("loading {} blocks", blockhashes.len());
                 self.req_send.send(Request::get_blocks(&blockhashes))
             })?;
 
-            for hash in blockhashes {
-                let block = self.blocks_duration.observe_duration("response", || {
-                    let block = self
-                        .blocks_recv
-                        .recv()
-                        .with_context(|| format!("failed to get block {}", hash))?;
-                    let header = bsl::BlockHeader::parse(&block[..])
-                        .expect("core returned invalid blockheader")
-                        .parsed_owned();
-                    ensure!(
-                        &header.block_hash_sha2()[..] == hash.as_byte_array(),
-                        "got unexpected block"
-                    );
-                    Ok(block)
-                })?;
-                self.blocks_duration
-                    .observe_duration("process", || func(hash, block));
-            }
-            Ok(())
+            rayon::scope(|s| {
+                let (send, receive) = std::sync::mpsc::channel();
+
+                for hash in blockhashes {
+                    let block = self.blocks_duration.observe_duration("response", || {
+                        let block = self
+                            .blocks_recv
+                            .recv()
+                            .with_context(|| format!("failed to get block {}", hash))?;
+                        let header = bsl::BlockHeader::parse(&block[..])
+                            .expect("core returned invalid blockheader")
+                            .parsed_owned();
+                        ensure!(
+                            &header.block_hash_sha2()[..] == hash.as_byte_array(),
+                            "got unexpected block"
+                        );
+                        Ok(block)
+                    })?;
+                    let func = &func;
+                    let blocks_duration = &self.blocks_duration;
+                    let send = send.clone();
+                    s.spawn(move |_| {
+                        let r = blocks_duration.observe_duration("process", || func(hash, block));
+                        let _ = send.send(r);
+                    });
+                }
+                let result: Result<Vec<_>, std::sync::mpsc::RecvError> =
```

Comment:
Is it OK to block here (from `rayon` point of view)?
Asking, since when using [this patch](https://github.com/romanz/electrs/commit/36c38284f989d3786e1e9e1bc560431bb1d946ea), it seems that a task is being spawned, but seems to not run:
```
[2023-11-20T19:44:54.785Z DEBUG electrs::p2p] peer inventory: [Block(0x423cfe9f541e1425b02b93d8db56f8fecdcb0ce8f58d33102d205c73f6bb037e)]
[2023-11-20T19:44:54.786Z DEBUG electrs::p2p] got 1 new headers
[2023-11-20T19:44:54.786Z INFO  electrs::index] indexing 1 blocks: [111..111]
[2023-11-20T19:44:54.786Z DEBUG electrs::p2p] loading 1 blocks
[2023-11-20T19:44:54.786Z DEBUG electrs::p2p] spawning task for 423cfe9f541e1425b02b93d8db56f8fecdcb0ce8f58d33102d205c73f6bb037e
[2023-11-20T19:44:54.786Z DEBUG electrs::p2p] waiting for 1 blocks
[2023-11-20T19:44:54.786Z DEBUG electrs::p2p] processing 423cfe9f541e1425b02b93d8db56f8fecdcb0ce8f58d33102d205c73f6bb037e
[2023-11-20T19:44:54.786Z DEBUG electrs::index] writing 3 funding and 11 spending rows from 2 transactions, 1 blocks
[2023-11-20T19:44:54.843Z INFO  electrs::signals] notified via SIG10
[2023-11-20T19:44:54.866Z INFO  electrs::chain] chain updated: tip=423cfe9f541e1425b02b93d8db56f8fecdcb0ce8f58d33102d205c73f6bb037e, height=111
[2023-11-20T19:44:54.866Z DEBUG electrs::p2p] loading 1 blocks
[2023-11-20T19:44:54.866Z DEBUG electrs::p2p] spawning task for 423cfe9f541e1425b02b93d8db56f8fecdcb0ce8f58d33102d205c73f6bb037e
[2023-11-20T19:44:54.866Z DEBUG electrs::p2p] waiting for 1 blocks
<it seems to get stuck here>
[2023-11-20T19:45:48.880Z DEBUG electrs::server] 0: recv {"jsonrpc":"2.0","method":"blockchain.estimatefee","id":283,"params":[25]}
[2023-11-20T19:45:48.880Z DEBUG electrs::server] 0: recv {"jsonrpc":"2.0","method":"blockchain.estimatefee","id":284,"params":[10]}
[2023-11-20T19:45:48.880Z DEBUG electrs::server] 0: recv {"jsonrpc":"2.0","method":"blockchain.estimatefee","id":285,"params":[5]}
[2023-11-20T19:45:48.880Z DEBUG electrs::server] 0: recv {"jsonrpc":"2.0","method":"blockchain.estimatefee","id":286,"params":[2]}
[2023-11-20T19:45:48.889Z DEBUG electrs::server] 0: recv {"jsonrpc":"2.0","method":"mempool.get_fee_histogram","id":287}
```
Not sure, but it may be related to https://github.com/rayon-rs/rayon/issues/835... WDYT?

Reply:
It seems a good catch! Did you have a chance to try the patch with `rayon::in_place_scope` instead of `rayon::scope` as suggested in the issue?

Another alternative maybe could be to use `Arc<Mutex<Vec<R>>>` instead of the channel (so the blocking happens inside the spawn)

Reply:
Unfortunately, I can't emulate the deadlock locally (I can see the output of another debug print after recv), maybe it depends on the number of threads used by rayon.
I am going to push a temporary commit to use the CI for this

Reply:
The CI is not happy with `in_place_scope` and neither with the Mutex version https://github.com/RCasatta/electrs/actions/runs/6941492681/job/18882464601

Reply:
> I can't emulate the deadlock locally (I can see the output of another debug print after recv)

With locally I mean without the docker...

How can I see the logs with the docker command as it is done (`docker run -v $PWD/contrib/:/contrib -v $PWD/tests/:/tests --rm electrs:tests bash /tests/run.sh`) in the CI  ?

Reply:
> Unfortunately, the CI is not happy with `in_place_scope` and neither with the Mutex version https://github.com/RCasatta/electrs/actions/runs/6941492681/job/18882464601

Please rebase over latest `master`, the above failure should be fixed by https://github.com/romanz/electrs/pull/939.

Reply:
> How can I see the logs with the docker command as it is done (`docker run -v $PWD/contrib/:/contrib -v $PWD/tests/:/tests --rm electrs:tests bash /tests/run.sh`) in the CI ?

electrs logs are stored under `/data/electrs` (inside the container):
https://github.com/romanz/electrs/blob/a01d58c35d2f6aaea7a479eb508404e7ff3affdf/tests/run.sh#L47

Reply:
Sorry for the dumb questions, but if the container dies after an error how to check inside the container?

Reply:
Sorry, the logs are indeed unavailable when the container fails in CI...

I have added two more commits (8326249088f251e92fd6621710144ea57338ea17 & acaebd0494e8392d8feadf7b85fd15f4dd1d9111), so it should be possible to reproduce the issue locally by:
```
$ docker build -f Dockerfile.ci . --rm -t electrs:tests-bug
$ docker run -v $PWD/contrib/:/contrib -v /tmp/data:/data -v $PWD/tests/:/tests --rm electrs:tests-bug bash -x /tests/run.sh
```
The logs should be under `/tmp/data/` after the container exits.

Reply:
Thanks, in the meantime I launched a couple of jobs on the CI.
The bad news is that the `in_place_scope` seems to deadlock the same https://github.com/RCasatta/electrs/actions/runs/6947160675
The good news is that the Mutex version isn't deadlocking https://github.com/RCasatta/electrs/actions/runs/6946996655
But before cleaning up the Mutex version I think it could be possible a mutex-free solution...

Reply:
> But before cleaning up the Mutex version I think it could be possible a mutex-free solution...

I was considering to use a `Vec<OnceLock<R>>` for thread results, however `OnceLock` has MSRV 1.70
So, I changed the `Mutex<Vec<R>>` to `Vec<Mutex<Option<R>>>` so that there are no blocking call excluding the end of the rayon scope